dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/TrackerObjectsMapper.java

```diff
@@ -0,0 +1,348 @@
+      @Nonnull org.hisp.dhis.tracker.imports.domain.TrackedEntity trackedEntity,
+      @Nonnull UserDetails user) {
+    TrackedEntity dbTrackedEntity = preheat.getTrackedEntity(trackedEntity.getTrackedEntity());
+    OrganisationUnit organisationUnit = preheat.getOrganisationUnit(trackedEntity.getOrgUnit());
+    TrackedEntityType trackedEntityType =
+        preheat.getTrackedEntityType(trackedEntity.getTrackedEntityType());
+
+    Date now = new Date();
+
+    if (dbTrackedEntity == null) {
+      dbTrackedEntity = new TrackedEntity();
+      dbTrackedEntity.setUid(trackedEntity.getTrackedEntity());
+      dbTrackedEntity.setCreated(now);
+      dbTrackedEntity.setCreatedByUserInfo(UserInfoSnapshot.from(user));
+      dbTrackedEntity.setStoredBy(trackedEntity.getStoredBy());
+    }
+
+    dbTrackedEntity.setLastUpdatedByUserInfo(UserInfoSnapshot.from(user));
+    dbTrackedEntity.setLastUpdated(now);
+    dbTrackedEntity.setPotentialDuplicate(trackedEntity.isPotentialDuplicate());
+    dbTrackedEntity.setCreatedAtClient(DateUtils.fromInstant(trackedEntity.getCreatedAtClient()));
+    dbTrackedEntity.setLastUpdatedAtClient(
+        DateUtils.fromInstant(trackedEntity.getUpdatedAtClient()));
+    dbTrackedEntity.setOrganisationUnit(organisationUnit);
+    dbTrackedEntity.setTrackedEntityType(trackedEntityType);
+    dbTrackedEntity.setInactive(trackedEntity.isInactive());
+    dbTrackedEntity.setGeometry(trackedEntity.getGeometry());
+
+    return dbTrackedEntity;
+  }
+
+  public static @Nonnull Enrollment map(
+      @Nonnull TrackerPreheat preheat,
+      @Nonnull org.hisp.dhis.tracker.imports.domain.Enrollment enrollment,
+      @Nonnull UserDetails user) {
+    Enrollment dbEnrollment = preheat.getEnrollment(enrollment.getEnrollment());
+    OrganisationUnit organisationUnit = preheat.getOrganisationUnit(enrollment.getOrgUnit());
+
+    Program program = preheat.getProgram(enrollment.getProgram());
+
+    TrackedEntity trackedEntity = preheat.getTrackedEntity(enrollment.getTrackedEntity());
+
+    Date now = new Date();
+
+    if (dbEnrollment == null) {
+      dbEnrollment = new Enrollment();
+      dbEnrollment.setUid(enrollment.getEnrollment());
+      dbEnrollment.setCreated(now);
+      dbEnrollment.setStoredBy(enrollment.getStoredBy());
+      dbEnrollment.setCreatedByUserInfo(UserInfoSnapshot.from(user));
+    }
+
+    dbEnrollment.setLastUpdated(now);
+    dbEnrollment.setLastUpdatedByUserInfo(UserInfoSnapshot.from(user));
+    dbEnrollment.setCreatedAtClient(DateUtils.fromInstant(enrollment.getCreatedAtClient()));
+    dbEnrollment.setLastUpdatedAtClient(DateUtils.fromInstant(enrollment.getUpdatedAtClient()));
+
+    Date enrollmentDate = DateUtils.fromInstant(enrollment.getEnrolledAt());
+    Date occurredDate = DateUtils.fromInstant(enrollment.getOccurredAt());
+
+    dbEnrollment.setEnrollmentDate(enrollmentDate);
+    dbEnrollment.setOccurredDate(occurredDate != null ? occurredDate : enrollmentDate);
```

Comment:
what's the difference between this one and `enrollmentDate`? and why do we set it to `enrollmentDate` if `occurredDate` is null?

Reply:
Now the there is less code and it is easier to understand this logic I created https://dhis2.atlassian.net/browse/DHIS2-18224 to reason about this kind of stuff.
I am not sure what is the logic here